crates/wasmtime/src/component/func.rs

```diff
@@ -241,4 +249,342 @@ impl Func {
+            // that this function itself needs to have its post-return called.
+            //
+            // The theory at least is that these two asserts ensure component
+            // model semantics are upheld where the host properly calls
+            // `post_return` on the right function despite the call being a
+            // separate step in the API.
+            assert!(
+                (*flags).needs_post_return(),
+                "post_return can only be called after a function has previously been called",
+            );
+            let post_return_arg = post_return_arg.expect("calling post_return on wrong function");
+
+            // This is a sanity-check assert which shouldn't ever trip.
+            assert!(!(*flags).may_enter());
+
+            // Unset the "needs post return" flag now that post-return is being
+            // processed. This will cause future invocations of this method to
+            // panic, even if the function call below traps.
+            (*flags).set_needs_post_return(false);
+
+            // If the function actually had a `post-return` configured in its
+            // canonical options that's executed here.
+            //
+            // Note that if this traps (returns an error) this function
+            // intentionally leaves the instance in a "poisoned" state where it
+            // can no longer be entered because `may_enter` is `false`.
+            if let Some((func, trampoline)) = post_return {
+                crate::Func::call_unchecked_raw(
+                    &mut store,
+                    func.anyfunc,
+                    trampoline,
+                    &post_return_arg as *const ValRaw as *mut ValRaw,
+                )?;
+            }
+
+            // And finally if everything completed successfully then the "may
+            // enter" flag is set to `true` again here which enables further use
+            // of the component.
+            (*flags).set_may_enter(true);
+        }
+        Ok(())
+    }
+
+    fn store_args<T>(
+        &self,
+        store: &mut StoreContextMut<'_, T>,
+        options: &Options,
+        params: &[Type],
+        args: &[Val],
+        dst: &mut MaybeUninit<[ValRaw; MAX_STACK_PARAMS]>,
+    ) -> Result<()> {
+        let mut size = 0;
+        let mut alignment = 1;
+        for ty in params {
+            alignment = alignment.max(ty.size_and_alignment().alignment);
+            ty.next_field(&mut size);
+        }
+
+        let mut memory = MemoryMut::new(store.as_context_mut(), options);
+        let ptr = memory.realloc(0, 0, alignment, size)?;
+        let mut offset = ptr;
+        for (ty, arg) in params.iter().zip(args) {
+            arg.store(&mut memory, ty.next_field(&mut offset))?;
+        }
+
+        map_maybe_uninit!(dst[0]).write(ValRaw::i64(ptr as i64));
+
+        Ok(())
+    }
+
+    fn load_result<'a>(
+        mem: &Memory,
+        ty: &Type,
+        src: &mut impl Iterator<Item = &'a ValRaw>,
```

Comment:
I worry with `impl Iterator` that this will excessively monomorphize into callers, so could this use something concrete like `&mut &[ValRaw]` or `&mut std::slice::Iter<'_, ValRaw>`?